storagemarket/impl/clientstates/client_fsm.go

```diff
@@ -0,0 +1,105 @@
+package clientstates
+
+import (
+	"github.com/filecoin-project/go-fil-markets/storagemarket"
+	"github.com/filecoin-project/go-statemachine/fsm"
+	"github.com/filecoin-project/specs-actors/actors/abi"
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+)
+
+// ClientEvents are the events that can happen in a storage client
+var ClientEvents = fsm.Events{
+	fsm.Event(storagemarket.ClientEventOpen).
+		From(storagemarket.StorageDealUnknown).To(nil),
+	fsm.Event(storagemarket.ClientEventEnsureFundsFailed).
+		From(storagemarket.StorageDealUnknown).To(storagemarket.StorageDealFailing).
+		Action(func(deal *storagemarket.ClientDeal, err error) error {
+			deal.Message = xerrors.Errorf("adding market funds failed: %w", err).Error()
+			return nil
+		}),
+	fsm.Event(storagemarket.ClientEventFundsEnsured).
+		From(storagemarket.StorageDealUnknown).To(storagemarket.StorageDealFundsEnsured),
+	fsm.Event(storagemarket.ClientEventWriteProposalFailed).
+		From(storagemarket.StorageDealFundsEnsured).To(storagemarket.StorageDealError).
+		Action(func(deal *storagemarket.ClientDeal, err error) error {
+			deal.Message = xerrors.Errorf("sending proposal to storage provider failed: %w", err).Error()
+			return nil
+		}),
+	fsm.Event(storagemarket.ClientEventDealProposed).
+		From(storagemarket.StorageDealFundsEnsured).To(storagemarket.StorageDealValidating),
```

Comment:
Ugh, I wonder what the race condition recovery looks like here with multiple simultaneous deals..

Reply:
well the upside is the provider validates it too... so maybe they'll catch it? But you are right that the logic ought to be ensure funds for all deals in progress... dunno what that looks like.